src/provider/hyrax_pc.rs

```diff
@@ -0,0 +1,414 @@
+    let R_size = (2usize).pow(right_num_vars as u32);
+    assert_eq!(L_size * R_size, n);
+
+    let comm = (0..L_size)
+      .collect::<Vec<usize>>()
+      .into_par_iter()
+      .map(|i| {
+        PedersenCommitmentEngine::commit(&ck.ck, &poly.get_Z()[R_size * i..R_size * (i + 1)])
+      })
+      .collect();
+
+    HyraxCommitment {
+      comm,
+      is_default: false,
+    }
+  }
+}
+
+impl<G: Group> TranscriptReprTrait<G> for HyraxCommitment<G> {
+  fn to_transcript_bytes(&self) -> Vec<u8> {
+    let mut v = Vec::new();
+    v.append(&mut b"poly_commitment_begin".to_vec());
+
+    for c in &self.comm {
+      v.append(&mut c.to_transcript_bytes());
+    }
+
+    v.append(&mut b"poly_commitment_end".to_vec());
+    v
+  }
+}
+
+impl<G: Group> TranscriptReprTrait<G> for HyraxCompressedCommitment<G> {
+  fn to_transcript_bytes(&self) -> Vec<u8> {
+    let mut v = Vec::new();
+    v.append(&mut b"poly_commitment_begin".to_vec());
+
+    for c in &self.comm {
+      v.append(&mut c.to_transcript_bytes());
+    }
+
+    v.append(&mut b"poly_commitment_end".to_vec());
+    v
+  }
+}
+
+/// Provides an implementation of the hyrax key
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct HyraxProverKey<G: Group> {
+  ck_s: CommitmentKey<G>,
+}
+
+/// Provides an implementation of the hyrax key
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct HyraxVerifierKey<G: Group> {
+  ck_v: CommitmentKey<G>,
+  ck_s: CommitmentKey<G>,
+}
+
+/// Provides an implementation of a polynomial evaluation argument
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct HyraxEvaluationArgument<G: Group> {
+  ipa: InnerProductArgument<G>,
+}
+
+/// Provides an implementation of a polynomial evaluation engine using Hyrax PC
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct HyraxEvaluationEngine<G: Group> {
+  _p: PhantomData<G>,
+}
+
+impl<G> EvaluationEngineTrait<G> for HyraxEvaluationEngine<G>
+where
+  G: Group<CE = HyraxCommitmentEngine<G>>,
+{
+  type CE = G::CE;
+  type ProverKey = HyraxProverKey<G>;
+  type VerifierKey = HyraxVerifierKey<G>;
+  type EvaluationArgument = HyraxEvaluationArgument<G>;
+
+  fn setup(
+    ck: &<Self::CE as CommitmentEngineTrait<G>>::CommitmentKey,
+  ) -> (Self::ProverKey, Self::VerifierKey) {
+    let pk = HyraxProverKey::<G> {
+      ck_s: G::CE::setup(b"hyrax", 1),
+    };
+
+    let vk = HyraxVerifierKey::<G> {
+      ck_v: ck.clone(),
+      ck_s: G::CE::setup(b"hyrax", 1),
+    };
+
+    (pk, vk)
+  }
+
+  fn prove(
+    ck: &CommitmentKey<G>,
+    pk: &Self::ProverKey,
+    transcript: &mut G::TE,
+    comm: &Commitment<G>,
+    poly: &[G::Scalar],
+    point: &[G::Scalar],
+    eval: &G::Scalar,
+  ) -> Result<Self::EvaluationArgument, SpartanError> {
+    transcript.absorb(b"poly_com", comm);
+
+    let poly_m = MultilinearPolynomial::<G::Scalar>::new(poly.to_vec());
```

Comment:
Same as above: if you're going to need to copy the full vector, you probably want to be explicit about it and make argument `poly: Vec<G::Scalar>`. ([C-CALLER-CONTROL](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control))

Take this with a grain of salt: `MultilinearPolynomial` in general is a bit unwieldy and would probably benefit from a larger refactoring based on `Iterator` APIs